### PR TITLE
feat(2467): [Breaking Change] Change log behave not to output log to file by default

### PR DESCRIPTION
**Pull request by yuichi10**
## Context

I checked how much of the sd-cmd log file is seen, and found that almost all people don't look at the log file.
Users spend time with artifact bookend teardown even if the user does not put files to artifacts because of sd-cmd.
It is better not to create log files to speed up build time.

## Objective

- Not to create log file as default.
- Able to output log files with `--debug` or `-debug` flag.
  - If the debug flag is true, all logs will be written to a file, but error logs will also be output to stderr.
- refactoring surrounded code that  I wrote.

## References

https://github.com/screwdriver-cd/screwdriver/issues/2467

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

**Comment by tk3fftk** [comment]
`gofmt` step has been failed due to try to use bash feature with sh. (It's not related to this fix but we should fix :pray:)

File: executor/executor_test.go [hunk 1]
```diff
@@ -164,49 +164,87 @@ func dummyCommandSpec(format string) (spec *util.CommandSpec) {
 	return spec
 }
 
+// TODO should not output log file without --log-file flag
```
Comment (tk3fftk) [inline executor/executor_test.go:167]: Are these still TODO?

File: executor/executor.go [hunk 2]
```diff
@@ -21,28 +23,49 @@ import (
 var (
 	command = exec.Command
 	lgr     *logger.Logger
+
+	hasOutputLogFile = false
```
Comment (tk3fftk) [inline executor/executor.go:27]: does `hasOutputLogFile`  have to be global?

Comment (yuichi10) [reply executor/executor.go:27]: I had made it a global variable because it would be easier to manage.
The [flag example](https://golang.org/pkg/flag/) is also set to global,  thus I thought this would be a good way to go.

File: executor/executor_test.go (hunk 1 above)
Comment (yuichi10) [reply executor/executor_test.go:167]: Thanks. I deleted

File: logger/logger.go [hunk 3]
```diff
@@ -4,35 +4,104 @@ package logger
...
+		l.hasOutputDebugLog = output
+		return nil
+	}
 }
 
 // New returns logger object
-func New(dirPath, filename string, flag int, debug bool) (*Logger, error) {
+func New(options ...LogOption) (*Logger, error) {
 	lgr := new(Logger)
+	lgr.errorFlag = log.LstdFlags
```
Comment (kkisic) [inline logger/logger.go:91]: Is it better to set the default for `lgr.debugFlag` here too?
Then the following functions could be made test-only functions.
- `func DebugFlag(flag int) LogOption`
- `func ErrorFlag(flag int) LogOption`

Comment (yuichi10) [reply logger/logger.go:91]: You are right. I fixed

File: logger/logger.go [hunk 4]
```diff
@@ -50,21 +113,31 @@ func CreateLogFile(dirPath, filename string) (io.WriteCloser, error) {
...
 
-// SetInfos set logger information from arguments
-func (l *Logger) SetInfos(file io.WriteCloser, flag int, debug bool) {
-	l.File = file
-	if debug {
-		l.Debug = log.New(io.MultiWriter(os.Stderr, file), "", flag)
-	} else {
-		l.Debug = log.New(file, "", flag)
+func (l *Logger) buildDebugLogger() {
+	if l.isDebug && l.file != nil {
```
Comment (kkisic) [inline logger/logger.go:117]: If the file exists only when `--debug` is used, I think then one of the conditions should be enough.

File: logger/logger.go (hunk 3 above)
Comment (yuichi10) [reply logger/logger.go:82]: Sorry, I forget the second opinion.
I think it's ok to put LogOption function to non-test code, because it it not strange that logger package have these Option.
Also, if I write these Option to test code, it is hard to set debug option from executor test.

File: logger/logger.go (hunk 4 above)
Comment (yuichi10) [reply logger/logger.go]: Exactly, I already forget to set isDebug = true.

File: logger/logger.go (hunk 3 above)
Comment (kkisic) [reply logger/logger.go:82]: Thx. I see.

**Comment by kumada626** [comment]
In the case that user want to output all cmd log in the build, users hard to add flag to all cmd in the build.
Another problem is that if one cmd is used by another cmd, adding a flag to the outer cmd will not add a flag to the inner cmd, and all logs will not be output.
How about making it possible to read flags from environment variables (such that `SD_CMD_DEBUG_LOG` ) as well?

**Comment by kkisic** [comment]
> In the case that user want to output all cmd log in the build, users hard to add flag to all cmd in the build.
> Another problem is that if one cmd is used by another cmd, adding a flag to the outer cmd will not add a flag to the inner cmd, and all logs will not be output.
> How about making it possible to read flags from environment variables (such that `SD_CMD_DEBUG_LOG` ) as well?

You're right.
I added it.

File: executor/executor.go [hunk 5]
```diff
@@ -112,3 +158,10 @@ func execCommand(path string, args []string) (err error) {
 	lgr.Debug.Printf("System Time: %v, User Time: %v\n", state.SystemTime(), state.UserTime())
 	return
 }
+
+// CleanUp close file you use.
+func CleanUp() {
```
Comment (yuichi10) [inline executor/executor.go:163]: Is it better to logger have a Close or Clean function?
```
package executor
lgr.Close() or lgr.Clean()

package logger
func (l *Logger)Close() {}
```

Comment (yuichi10) [reply executor/executor.go:163]: Or when nil writeCloser is comming, logger set ioutil.Discard.
Then executor do not need check file is nil or not

File: executor/executor_test.go [hunk 6]
```diff
@@ -165,49 +170,97 @@ func dummyCommandSpec(format string) (spec *util.CommandSpec) {
...
 
 func TestNew(t *testing.T) {
-	// case binary format
-	spec := dummyCommandSpec(binaryFormat)
-	sdapi := newDummySDAPI(spec, nil)
-	// success
-	executor, err := New(sdapi, []string{"ns/cmd@ver"})
-	if err != nil {
-		t.Errorf("err=%q, want nil", err)
+	successCases := []struct {
```
Comment (yuichi10) [inline executor/executor_test.go:173]: Do we need to add a test case that check  the file is not empty, config.Debug = true.

Comment (kkisic) [reply executor/executor_test.go:173]: Thx. I added it.

File: executor/executor.go (hunk 5 above)
Comment (kkisic) [reply executor/executor.go:163]: I fixed it.
I've decided to let logger have Close function.

**Review by yuichi10 — COMMENTED** [review]
LGTM

**Review by kumada626 — APPROVED** [review]

**Comment by sugarnaoming** [comment]
I think it would be useful to write down what options are available in README.

**Comment by kkisic** [comment]
> I think it would be useful to write down what options are available in README.

Thx.
I added a note about it in the README.

**Review by sugarnaoming — APPROVED** [review]

**Review by jithine — APPROVED** [review]

**Review by tkyi — APPROVED** [review]

**Review by pritamstyz4ever — APPROVED** [review]
